File: 351004/Dashkevich/Kafka_lab/publisher/src/main/java/com/adashkevich/kafka/lab/service/MarkerService.java
package com.adashkevich.kafka.lab.service;

import com.adashkevich.kafka.lab.dto.request.MarkerRequestTo;
import com.adashkevich.kafka.lab.dto.response.MarkerResponseTo;
import com.adashkevich.kafka.lab.exception.ConflictException;
import com.adashkevich.kafka.lab.exception.NotFoundException;
import com.adashkevich.kafka.lab.model.News;
import com.adashkevich.kafka.lab.repository.MarkerRepository;
import com.adashkevich.kafka.lab.repository.NewsRepository;
import com.adashkevich.kafka.lab.mapper.MarkerMapper;
import com.adashkevich.kafka.lab.model.Marker;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional(readOnly = true)
public class MarkerService {
    private final MarkerRepository repo;
    private final NewsRepository newsRepo;
    private final MarkerMapper mapper;

    public MarkerService(MarkerRepository repo, NewsRepository newsRepo, MarkerMapper mapper) {
        this.repo = repo;
        this.newsRepo = newsRepo;
        this.mapper = mapper;
    }

    @Transactional
    public MarkerResponseTo create(MarkerRequestTo dto) {
        ensureNameUnique(dto.name, null);
        Marker saved = repo.save(mapper.toEntity(dto));
        return mapper.toResponse(saved);
    }

    public List<MarkerResponseTo> getAll() {
        return repo.findAll(Sort.by("id")).stream().map(mapper::toResponse).toList();
    }

    public MarkerResponseTo getById(Long id) {
        return mapper.toResponse(findExisting(id));
    }

    @Transactional
    public MarkerResponseTo update(Long id, MarkerRequestTo dto) {
        Marker existing = findExisting(id);
        ensureNameUnique(dto.name, id);
        existing.setName(dto.name);
        return mapper.toResponse(repo.save(existing));
    }

    @Transactional
    public void delete(Long id) {
        Marker marker = findExisting(id);
        for (News news : newsRepo.findAll()) news.getMarkers().removeIf(m -> id.equals(m.getId()));
        repo.delete(marker);
    }

    private Marker findExisting(Long id) {
        return repo.findById(id).orElseThrow(() -> new NotFoundException("Marker not found", "40430"));
    }

    private void ensureNameUnique(String name, Long selfId) {
        boolean exists = selfId == null ? repo.existsByNameIgnoreCase(name) : repo.existsByNameIgnoreCaseAndIdNot(name, selfId);
        if (exists) throw new ConflictException("Marker name must be unique", "40930");
    }
}

File: 351002/Mikhalkov/publisher/Distcomp.Publisher.Application/Services/MarkerService.cs
using AutoMapper;
using Distcomp.Application.DTOs;
using Distcomp.Application.Exceptions;
using Distcomp.Application.Interfaces;
using Distcomp.Domain.Models;

namespace Distcomp.Application.Services
{
    public class MarkerService : IMarkerService
    {
        private readonly IRepository<Marker> _repository;
        private readonly IMapper _mapper;

        public MarkerService(IRepository<Marker> repository, IMapper mapper)
        {
            _repository = repository;
            _mapper = mapper;
        }

        public MarkerResponseTo Create(MarkerRequestTo request)
        {
            ValidateRequest(request);

            if (_repository.GetAll().Any(m => m.Name.Equals(request.Name, StringComparison.OrdinalIgnoreCase)))
                throw new RestException(403, 40302, $"Marker with name '{request.Name}' already exists");

            var marker = _mapper.Map<Marker>(request);
            var created = _repository.Create(marker);
            return _mapper.Map<MarkerResponseTo>(created);
        }

        public MarkerResponseTo? GetById(long id)
        {
            var marker = _repository.GetById(id);
            if (marker == null) throw new RestException(404, 40403, "Marker not found");
            return _mapper.Map<MarkerResponseTo>(marker);
        }

        public IEnumerable<MarkerResponseTo> GetAll() =>
            _mapper.Map<IEnumerable<MarkerResponseTo>>(_repository.GetAll());

        public MarkerResponseTo Update(long id, MarkerRequestTo request)
        {
            var existing = _repository.GetById(id);
            if (existing == null) throw new RestException(404, 40403, "Cannot update: Marker not found");

            ValidateRequest(request);

            if (!existing.Name.Equals(request.Name, StringComparison.OrdinalIgnoreCase) &&
                _repository.GetAll().Any(m => m.Name.Equals(request.Name, StringComparison.OrdinalIgnoreCase)))
                throw new RestException(403, 40302, "New marker name is already taken");

            _mapper.Map(request, existing);
            existing.Id = id;
            _repository.Update(existing);

            return _mapper.Map<MarkerResponseTo>(existing);
        }

        public bool Delete(long id)
        {
            if (_repository.GetById(id) == null) throw new RestException(404, 40403, "Cannot delete: Marker not found");
            return _repository.Delete(id);
        }

        private void ValidateRequest(MarkerRequestTo request)
        {
            if (request.Name.Length < 2 || request.Name.Length > 32)
                throw new RestException(400, 40007, "Marker name must be between 2 and 32 characters");
        }
    }
}